core/src/main/java/io/grpc/ClientInterceptor.java

```diff
@@ -31,35 +31,74 @@
 
 package io.grpc;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+
 import javax.annotation.concurrent.ThreadSafe;
 
 /**
- * Interface for intercepting outgoing calls before they are dispatched by a {@link Channel}.
+ * A {@link CallFactory} for intercepting outgoing calls before they are dispatched by a {@link
+ * CallFactory}.
  *
- * <p>Implementers use this mechanism to add cross-cutting behavior to {@link Channel} and
- * stub implementations. Common examples of such behavior include:
- * <ul>
- * <li>Adding credentials to header metadata</li>
- * <li>Logging and monitoring call behavior</li>
- * <li>Request and response rewriting</li>
- * </ul>
+ * <p>Implementers use this mechanism to add cross-cutting behavior to {@link CallFactory} and stub
+ * implementations. Common examples of such behavior include: <ul> <li>Adding credentials to header
+ * metadata</li> <li>Logging and monitoring call behavior</li> <li>Request and response
+ * rewriting</li> </ul>
  */
 @ThreadSafe
-public interface ClientInterceptor {
+public abstract class ClientInterceptor extends CallFactory {
+  private CallFactory next;
```

Comment:
I feel the design to make the interface a stateful abstract class is a bit concerning. In many cases, an interceptor chain can be created as an immutable object.

Two alternatives:
- keep the old design and drop implementation of CallFactory
- strengthen the contract and introduce getNext as abstract/interface method. Then subclasses decide how to create and store chaining.